src/core/common/instance.cpp

```diff
@@ -240,6 +240,16 @@ Instance::Instance(void)
 {
 }
 
+#if OPENTHREAD_MTD || OPENTHREAD_FTD
+#if !OPENTHREAD_CONFIG_HEAP_EXTERNAL_ENABLE
```

Comment:
How about combining the `#if` into one?
```suggestion
#if (OPENTHREAD_MTD || OPENTHREAD_FTD) && !OPENTHREAD_CONFIG_HEAP_EXTERNAL_ENABLE
```